server/coinlock/coinlocker.go

```diff
@@ -0,0 +1,222 @@
+func NewMasterCoinLocker() *MasterCoinLocker {
+	return &MasterCoinLocker{
+		bookLock: NewAssetCoinLocker(),
+		swapLock: NewAssetCoinLocker(),
+	}
+}
+
+// CoinLocked indicates if a coin is locked by either the swap or book lock.
+func (cl *MasterCoinLocker) CoinLocked(coin CoinID) bool {
+	lockedBySwap := cl.swapLock.CoinLocked(coin)
+	if lockedBySwap {
+		return true
+	}
+
+	return cl.bookLock.CoinLocked(coin)
+}
+
+// OrderCoinsLocked lists all coins locked by a given order are locked by either
+// the swap or book lock.
+func (cl *MasterCoinLocker) OrderCoinsLocked(oid order.OrderID) []CoinID {
+	coins := cl.swapLock.OrderCoinsLocked(oid)
+	if len(coins) > 0 {
+		return coins
+	}
+
+	// TODO: merge?
+	return cl.bookLock.OrderCoinsLocked(oid)
+}
+
+// Book provides the market-level CoinLocker.
+func (cl *MasterCoinLocker) Book() CoinLocker {
+	return &bookLocker{cl}
+}
+
+// Swap provides the swap-level CoinLocker.
+func (cl *MasterCoinLocker) Swap() CoinLocker {
+	return &swapLocker{cl}
+}
+
+type bookLocker struct {
+	*MasterCoinLocker
+}
+
+func (bl *bookLocker) LockOrdersCoins(orders []order.Order) {
+	bl.bookLock.LockOrdersCoins(orders)
+}
+
+func (bl *bookLocker) LockCoins(orderCoins map[order.OrderID][]CoinID) {
+	bl.bookLock.LockCoins(orderCoins)
+}
+
+func (bl *bookLocker) UnlockOrderCoins(oid order.OrderID) {
+	bl.bookLock.UnlockOrderCoins(oid)
+}
+
+var _ (CoinLocker) = (*bookLocker)(nil)
+
+type swapLocker struct {
+	*MasterCoinLocker
+}
+
+func (bl *swapLocker) LockOrdersCoins(orders []order.Order) {
+	bl.swapLock.LockOrdersCoins(orders)
+}
+
+func (bl *swapLocker) LockCoins(orderCoins map[order.OrderID][]CoinID) {
+	bl.swapLock.LockCoins(orderCoins)
+}
+
+func (bl *swapLocker) UnlockOrderCoins(oid order.OrderID) {
+	bl.swapLock.UnlockOrderCoins(oid)
+}
+
+var _ (CoinLocker) = (*swapLocker)(nil)
+
+type coinIDKey string
+
+// AssetCoinLocker is a coin locker for a single asset. Do not use this for more
+// than one asset.
+type AssetCoinLocker struct {
+	coinMtx            sync.RWMutex
+	lockedCoins        map[coinIDKey]struct{}
+	lockedCoinsByOrder map[order.OrderID][]CoinID
+}
+
+// NewAssetCoinLocker constructs a new AssetCoinLocker.
+func NewAssetCoinLocker() *AssetCoinLocker {
+	return &AssetCoinLocker{
+		lockedCoins:        make(map[coinIDKey]struct{}),
+		lockedCoinsByOrder: make(map[order.OrderID][]CoinID),
+	}
+}
+
+// CoinLocked indicates if a coin identifier (e.g. UTXO) is locked.
+func (ac *AssetCoinLocker) CoinLocked(coin CoinID) bool {
+	ac.coinMtx.RLock()
+	_, locked := ac.lockedCoins[coinIDKey(coin)]
+	ac.coinMtx.RUnlock()
+	return locked
+}
+
+// OrderCoinsLocked lists the coin IDs (e.g. UTXOs) locked by an order.
+func (ac *AssetCoinLocker) OrderCoinsLocked(oid order.OrderID) []CoinID {
+	ac.coinMtx.RLock()
+	defer ac.coinMtx.RUnlock()
+	return ac.lockedCoinsByOrder[oid]
+}
+
+// UnlockOrderCoins unlocks any coins backing order.
+func (ac *AssetCoinLocker) UnlockOrderCoins(oid order.OrderID) {
```

Comment:
Unlocking by order ID may prove to be in appropriate for the same reasons mentioned before.  Unlocking by CoinID may be needed.

Reply:
May prove to be a nice feature, but not a priority, IMO. Locking only the coins that have been spent is possible with no changes to the current backend interface though.